[test] Simplify `st.audio_input` test suite

sfc-gh-nbellante opened the pull request:
## Describe your changes

Refactored the audio input component tests to improve reliability and maintainability. 

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante commented:
* **#12539** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12539?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12542** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12542?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12542?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12541** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12541?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12540** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12540?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sfc-gh-bnisco commented on e2e_playwright/st_audio_input_test.py:
```diff
@@ -15,581 +15,395 @@
     get_element_by_key,
 )
 
+# Audio input indices in test app
```
**suggestion:** These should be using accessible-based accessors rather than index-based accessors.

While in the past we've used index-based accessors in e2e tests, switching the order around can cause unnecessary noise. Even more importantly, is that index-based accessors induce a specific runtime ordering requirement, which can cause flakiness. In order to combat both the tech debt and uplevel our a11y in app, we've been moving to accessible selectors that select based on things like label. Some example PRs: https://github.com/streamlit/streamlit/pull/12414, https://github.com/streamlit/streamlit/pull/12362, https://github.com/streamlit/streamlit/pull/12371

sfc-gh-nbellante replied:
For sure, I'll be switching this in the next PR. I wanted to avoid conflicting back with the other PR that simplifies the test file.

sfc-gh-bnisco commented on e2e_playwright/st_audio_input_test.py:
```diff
@@ -15,581 +15,395 @@
-    expect(time_code).to_be_visible()
-    expect(time_code).not_to_have_text("00:00")
+    # Record button should be visible (allows re-recording)
+    record_button = audio_input.locator('[aria-label="Record"]')
```
**suggestion:** Prefer to utilize built-in selectors: `audio_input.get_by_role("button", name="Record")`. This feedback applies to all `locator('[aria-label=""]')` calls in this PR

sfc-gh-nbellante committed:
[test] Simplify and refactor audio input test suite

- Add constants for all audio input indices for better readability
- Create comprehensive verification helper functions (verify_recording_exists, verify_no_recording, verify_recording_in_progress)
- Add grant_microphone_permissions helper to handle webkit gracefully
- Refactor record_and_stop helper to be more flexible
- Add test_help_tooltip to verify help parameter functionality
- Use helper functions consistently across all tests
- Fix test to expect 12 audio inputs instead of 11
- Improve code organization with cleaner helper functions

sfc-gh-nbellante committed:
snapshots

sfc-gh-nbellante committed:
Fix strict mode violations in audio input tests

Use exact=True parameter in get_by_role selectors to avoid ambiguity
between Record and Clear recording buttons
